### PR TITLE
Support asynchronous tests

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit adds support for executing tests asynchronously. This is modeled
by tests returning a `Future` instead of simply executing inline, and is
signified with `#[wasm_bindgen_test(async)]`.

Support for this is added through a new `wasm-bindgen-futures` crate which is a
binding between the `futures` crate and JS `Promise` objects.

Lots more details can be found in the details of the commit, but one of the end
results is that the `web-sys` tests are now entirely contained in the same test
suite and don't need `npm install` to be run to execute them!

### CODE REVIEW

**Comment by sendilkumarn** [comment]
Awesome ❤️ it

File: crates/futures/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+/// # Panics
+///
+/// Note that in wasm panics are currently translated to aborts, but "abort" in
+/// this case means that a JS exception is thrown. The wasm module is still
+/// usable (likely erroneously) after Rust panics.
+///
+/// If the `future` provided panics then the returned `Promise` **will not
+/// resolve**. Instead it will be a leaked promise. This is an unfortunate
+/// limitation of wasm currently that's hoped to be fixed one day!
+pub fn rust2js<F>(future: F) -> Promise
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/futures/src/lib.rs]: Heh you're not kidding, `rust2js` is a bit of a random name. I'll rename!

File: src/closure.rs [hunk 2]
```diff
@@ -65,7 +69,9 @@ use describe::*;
 /// }
 /// ```
 pub struct Closure<T: ?Sized> {
-    inner: UnsafeCell<Box<T>>,
+    // Actuall a `Rc` pointer, but in raw form so we can easily make copies.
+    // See below documentation for why this is in an `Rc`.
+    inner: *const UnsafeCell<Box<T>>,
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply src/closure.rs]: The purpose here was for the `IntoWasmAbi for &'a Closure` below where we need to pull out the raw pointer and ideally would avoid a `clone` + `drop`. It looks like it consumes self, but not quite :(

File: crates/futures/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+                    Err(value) => (value, &me.reject),
+
+                    // Otherwise keep going in our loop, if we weren't notified
+                    // we'll break out and start waiting.
+                    Ok(Async::NotReady) => continue,
+                };
+
+                let array = Array::new();
+                array.push(&val);
+                drop(f.apply(&JsValue::undefined(), &array));
```
[the remark this answers is not in the document]
Comment (alexcrichton) [reply crates/futures/src/lib.rs]: Excellent idea, done now!

**Pull request merged**